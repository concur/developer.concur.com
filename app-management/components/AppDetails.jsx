/* eslint-env browser */

import React, { PropTypes } from 'react';
import { Link } from 'react-router';

class AppDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      app: {},
      loading: false,
      error: false,
    };
  }

  componentWillMount() {
    window.fetch(`${process.env.API_SERVER}/apps/${this.props.params.id}`)
      .then(response => response.json())
      .then(app => this.setState({
        app,
        loading: false,
      }))
      .catch(() => this.setState({
        error: true,
        loading: false,
      }));
  }

  render () {
    if (this.state.loading) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>Loading...</p>
          </div>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>An error occurred when loading this app</p>
            <Link to="/">Back to My Apps</Link>
          </div>
        </div>
      );
    }

    const { app } = this.state;

    return (
      <div className="row">
        <div className="col-md-12">
          <h2>{app.name}</h2>
          <Link to="/">&larr; Back to My Apps</Link>
          <div className="row">
            <p>{app.description}</p>
          </div>
        </div>
      </div>
    );
  }
}

AppDetails.propTypes = {
  params: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

export default AppDetails;
