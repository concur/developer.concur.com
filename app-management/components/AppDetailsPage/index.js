import { connect } from 'react-redux';
import AppDetailsPage from './AppDetailsPage';
import { fetchAppDetails, updateAppDetails } from '../../actions/appDetails';

export function mapStateToProps(state) {
  return state.appDetails;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppDetails: id => dispatch(fetchAppDetails(id)),
    handleSubmit: app => dispatch(updateAppDetails(app)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppDetailsPage);
