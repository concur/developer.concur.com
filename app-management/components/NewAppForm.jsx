import React from 'react';

import grants from '../data/grants.json';
import scopes from '../data/scopes.json';

class NewAppForm extends React.Component {
  render () {
    const grantOptions = grants.map((grant, idx) => {
      const { title, value } = grant;
      return <option key={idx} value={value}>{title}</option>;
    });
    const scopeOptions = scopes.map((scope, idx) => {
      const { title, value } = scope;
      return <option key={idx} value={value}>{title}</option>;
    });

    return (
      <div className="row">
        <div className="col-md-12">
          <h2>App Creation Form</h2>
          <form
            action="{{ site.data.forms.server }}/apps"
            method="POST"
            id="appCreation"
            className="sky-form"
          >
            <fieldset>
              <div className="row">
                <section className="col col-6">
                  <label className="label" htmlFor="appName">App Name &#42;</label>
                  <div className="input">
                    <input type="text" name="appName" id="appName" placeholder="App Name" />
                  </div>
                </section>
                <section className="col col-6">
                  <label className="label" htmlFor="appDescription">App Description &#42;</label>
                  <div className="textarea">
                    <textarea
                      name="appDescription"
                      id="appDescription"
                      placeholder="App Description"
                    />
                  </div>
                </section>
              </div>
              <div className="row">
                <section className="col col-6">
                  <label className="label" htmlFor="applicationType">Application Type &#42;</label>
                  <div className="select">
                    <select name="appType" id="appType">
                      <option value="" defaultValue />
                      <option value="business">Business</option>
                      <option value="consumer">Consumer</option>
                    </select>
                    <i />
                  </div>
                </section>
                <section className="col col-6">
                  <label className="label" htmlFor="redirectUrl">Redirect URL &#42;</label>
                  <div className="input">
                    <input
                      type="url"
                      name="redirectUrl"
                      id="redirectUrl"
                      placeholder="Redirect URL"
                    />
                  </div>
                </section>
              </div>
              <div className="row">
                <section className="col col-6">
                  <label className="label" htmlFor="allowedGrants">Allowed Grants &#42;</label>
                  <div className="select">
                    <select name="allowedGrants" id="allowedGrants">
                      <option value="" defaultValue />
                      {grantOptions}
                    </select>
                    <i />
                  </div>
                </section>
                <section className="col col-6">
                  <label className="label" htmlFor="allowedScopes">Allowed Scopes &#42;</label>
                  <div className="select select-multiple">
                    <select name="allowedScopes" id="allowedScopes" multiple>
                      {scopeOptions}
                    </select>
                  </div>
                </section>
              </div>
              <section>
                <label className="checkbox" htmlFor="termsOfUseAgreement">
                  <input type="checkbox" name="termsOfUseAgreement" id="termsOfUseAgreement" />
                  <i />
                  I Agree to the <a href="/Terms-of-Use.html" target="_blank" rel="noopener noreferrer">Terms of Use</a> and <a href="https://www.concur.com/en-us/privacy-policy" target="_blank" rel="noopener noreferrer">Privacy Policy</a>
                </label>
              </section>
              <p>&#42; required field</p>
            </fieldset>
            <footer>
              <button type="submit" className="button">Submit</button>
            </footer>
          </form>
        </div>
      </div>
    );
  }
}

export default NewAppForm;
