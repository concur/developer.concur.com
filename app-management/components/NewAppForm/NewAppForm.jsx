import React, { PropTypes } from 'react';

import grants from '../../data/grants.json';
import scopes from '../../data/scopes.json';

const NewAppForm = (props) => {
  const grantOptions = grants.map((grant, idx) => {
    const { title, value } = grant;
    return <option key={idx} value={value}>{title}</option>;
  });
  const scopeOptions = scopes.map((scope, idx) => {
    const { title, value } = scope;
    return <option key={idx} value={value}>{title}</option>;
  });

  return (
    <div className="row">
      <div className="col-md-12">
        <h2>App Creation Form</h2>
        <form
          className="sky-form"
          method="POST"
          action="{{ site.data.forms.server }}/apps"
          onSubmit={props.postNewApp}
        >
          <fieldset>
            <div className="row">
              <section className="col col-6">
                <label className="label" htmlFor="appName">App Name &#42;</label>
                <div className="input">
                  <input
                    type="text"
                    name="appName"
                    id="appName"
                    placeholder="App Name"
                    value={props.appName}
                    onChange={props.handleInputChange}
                  />
                </div>
              </section>
              <section className="col col-6">
                <label className="label" htmlFor="appDescription">App Description &#42;</label>
                <div className="textarea">
                  <textarea
                    name="appDescription"
                    id="appDescription"
                    placeholder="App Description"
                    value={props.appDescription}
                    onChange={props.handleInputChange}
                  />
                </div>
              </section>
            </div>
            <div className="row">
              <section className="col col-6">
                <label className="label" htmlFor="appType">Application Type &#42;</label>
                <div className="select">
                  <select
                    name="appType"
                    id="appType"
                    value={props.appType}
                    onChange={props.handleInputChange}
                  >
                    <option value="" defaultValue />
                    <option value="business">Business</option>
                    <option value="consumer">Consumer</option>
                  </select>
                  <i />
                </div>
              </section>
              <section className="col col-6">
                <label className="label" htmlFor="redirectUrl">Redirect URL &#42;</label>
                <div className="input">
                  <input
                    type="url"
                    name="redirectUrl"
                    id="redirectUrl"
                    placeholder="Redirect URL"
                    value={props.redirectUrl}
                    onChange={props.handleInputChange}
                  />
                </div>
              </section>
            </div>
            <div className="row">
              <section className="col col-6">
                <label className="label" htmlFor="allowedGrants">Allowed Grants &#42;</label>
                <div className="select select-multiple">
                  <select
                    name="allowedGrants"
                    id="allowedGrants"
                    multiple
                    value={props.allowedGrants}
                    onChange={props.handleMultipleSelectChange}
                  >
                    {grantOptions}
                  </select>
                </div>
              </section>
              <section className="col col-6">
                <label className="label" htmlFor="allowedScopes">Allowed Scopes &#42;</label>
                <div className="select select-multiple">
                  <select
                    name="allowedScopes"
                    id="allowedScopes"
                    multiple
                    value={props.allowedScopes}
                    onChange={props.handleMultipleSelectChange}
                  >
                    {scopeOptions}
                  </select>
                </div>
              </section>
            </div>
            <section>
              <label className="checkbox" htmlFor="termsOfUseAgreement">
                <input
                  type="checkbox"
                  name="termsOfUseAgreement"
                  id="termsOfUseAgreement"
                  value={props.termsOfUseAgreement}
                  onChange={props.handleCheckboxChange}
                />
                <i />
                I Agree to the <a href="/Terms-of-Use.html" target="_blank" rel="noopener noreferrer">Terms of Use</a> and <a href="https://www.concur.com/en-us/privacy-policy" target="_blank" rel="noopener noreferrer">Privacy Policy</a>
              </label>
            </section>
            <p>&#42; required field</p>
          </fieldset>
          <footer>
            <button type="submit" className="button">Submit</button>
          </footer>
        </form>
      </div>
    </div>
  );
};

NewAppForm.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  handleCheckboxChange: PropTypes.func.isRequired,
  handleMultipleSelectChange: PropTypes.func.isRequired,
  postNewApp: PropTypes.func.isRequired,
  appName: PropTypes.string.isRequired,
  appDescription: PropTypes.string.isRequired,
  appType: PropTypes.string.isRequired,
  redirectUrl: PropTypes.string.isRequired,
  allowedGrants: PropTypes.arrayOf(PropTypes.string).isRequired,
  allowedScopes: PropTypes.arrayOf(PropTypes.string).isRequired,
  termsOfUseAgreement: PropTypes.bool.isRequired,
};

export default NewAppForm;
