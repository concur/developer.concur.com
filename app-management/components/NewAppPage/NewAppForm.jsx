import React, { PropTypes } from 'react';
import { Field, FieldArray, reduxForm } from 'redux-form';

import { formValidator } from '../../utils/formValidator';
import {
  InputField,
  TextareaField,
  CheckboxField,
  MultiselectField,
  RedirectUris,
  FieldHelp,
} from '../FormFields';

// All selectable grants and scopes
import grants from '../../data/grants.json';
import scopes from '../../data/scopes.json';
import appTypes from '../../data/appTypes.json';

const constraints = {
  name: {
    presence: { message: 'is required' },
    length: {
      minimum: 3,
      maximum: 99,
    },
  },
  description: {
    presence: { message: 'is required' },
    length: {
      minimum: 10,
      maximum: 4999,
    },
  },
  appType: {
    presence: { message: 'is required' },
  },
  redirectUris: {
    validateUrlArray: {
      allowLocal: true,
    },
    firstElementRequired: false,
  },
  allowedGrants: {
    presence: { message: '- at least one is required.' },
  },
  allowedScopes: {
    presence: { message: '- at least one is required.' },
  },
  termsOfUseAgreement: {
    presence: { message: 'is required' },
    inclusion: {
      within: [true],
      message: 'is required',
    },
  },
};

const defaultGrants = [grants.find(grant => grant.value === 'refresh_token')];
const defaultScopes = [scopes.find(scope => scope.value === 'openid')];
const defaultUris = [''];

const NewAppForm = ({ handleSubmit }) => (
  <form onSubmit={handleSubmit}>
    <p>&#42; required field</p>
    <fieldset>
      <div className="row">
        <section className="col-md-8">
          <Field
            component={InputField}
            type="text"
            name="name"
            label="App Name &#42;"
            placeholder="App Name"
          />
        </section>
        <section className="col-md-8">
          <Field
            component={TextareaField}
            name="description"
            label="App Description &#42;"
            placeholder="App Description"
          />
        </section>
        <section className="col-md-8">
          <Field
            component={MultiselectField}
            type="select-multiple"
            name="appType"
            label="App Type &#42;"
            options={appTypes}
          />
        </section>
        <section className="col-md-8">
          <FieldArray component={RedirectUris} name="redirectUris" />
          <FieldHelp>
            Valid redirect URIs that can be used for the Authorization Code grant type.
            <a
              href="https://tools.ietf.org/html/rfc6749#section-4.1"
              target="_blank"
              rel="noopener noreferrer"
            > See RFC6749
            </a>
          </FieldHelp>
        </section>
        <section className="col-md-8">
          <Field
            component={MultiselectField}
            type="select-multiple"
            name="allowedGrants"
            label="Allowed Grants &#42;"
            options={grants}
          />
          <FieldHelp>
            <a
              href="/api-reference/authentication/apidoc.html"
              target="_blank"
              rel="noopener noreferrer"
            >
              Authentication Grant documentation
            </a>
          </FieldHelp>
        </section>
        <section className="col-md-8">
          <Field
            component={MultiselectField}
            type="select-multiple"
            name="allowedScopes"
            label="Allowed Scopes &#42;"
            options={scopes}
          />
          <FieldHelp>
            <a
              href="/api-reference/authentication/scopes.html"
              target="_blank"
              rel="noopener noreferrer"
            >
              Scope documentation
            </a>
          </FieldHelp>
        </section>
      </div>
      <section>
        <Field
          component={CheckboxField}
          name="termsOfUseAgreement"
        >
          I Agree to the <a href="/Terms-of-Use.html" target="_blank" rel="noopener noreferrer">Terms of Use</a> and <a href="https://www.concur.com/en-us/privacy-policy" target="_blank" rel="noopener noreferrer">Privacy Policy</a>
        </Field>
      </section>
    </fieldset>
    <footer>
      <button type="submit" className="btn bright-blue">Submit</button>
    </footer>
  </form>
);

NewAppForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};

export default reduxForm({
  form: 'newApp',
  validate: formValidator(constraints),
  initialValues: {
    redirectUris: defaultUris,
    allowedGrants: defaultGrants,
    allowedScopes: defaultScopes,
  },
})(NewAppForm);
