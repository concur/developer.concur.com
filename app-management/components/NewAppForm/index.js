import { connect } from 'react-redux';
import NewAppForm from './NewAppForm';
import {
  handleInputChange, handleCheckboxChange,
  handleMultipleSelectChange, postNewApp,
} from '../../actions/newApp';

function mapStateToProps(state) {
  return state.newAppForm;
}

function mapDispatchToProps(dispatch) {
  return {
    handleInputChange: e => dispatch(handleInputChange(e.target.name, e.target.value)),
    handleCheckboxChange: e => dispatch(handleCheckboxChange(e.target.name)),
    handleMultipleSelectChange: e => dispatch(handleMultipleSelectChange(
      e.target.name,
      e.target.options,
    )),
    postNewApp: (e) => {
      e.preventDefault();
      dispatch(postNewApp());
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewAppForm);
