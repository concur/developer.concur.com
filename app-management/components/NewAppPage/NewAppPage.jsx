import React, { PropTypes } from 'react';
import NewAppForm from './NewAppForm';
import LoadingSpinner from '../LoadingSpinner';
import ErrorAlert from '../ErrorAlert';

const NewAppPage = ({ handleSubmit, isFetching, error }) => (
  <div className="row">
    <div className="col-md-12">
      <h2>Create an App</h2>
      <LoadingSpinner loading={isFetching} message="Submitting..." />
      <ErrorAlert error={error} />
      <div className="alert alert-info">
        <p>
          Register your new application (app) with Concur by filling out this form.
          Once you have registered an app, you will receive a clientId and clientSecret.
          The clientId is a unique UUID4 identifier for your app, and the clientSecret is
          your appâ€™s password. You will be using this credential to obtain tokens either
          for the app itself, or on behalf of a user.
        </p>
        <p>
          For more information on this process, please refer to our Getting Started guide for
          <a
            target="_blank"
            rel="noopener noreferrer"
            href="/api-reference/authentication/getting-started.html"
          > authentication
          </a>
        </p>
      </div>
      <NewAppForm onSubmit={handleSubmit} />
    </div>
  </div>
);

NewAppPage.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
};

export default NewAppPage;
