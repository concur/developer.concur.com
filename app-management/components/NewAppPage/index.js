import { connect } from 'react-redux';
import NewAppPage from './NewAppPage';
import { postNewApp } from '../../actions/newApp';

export function mapStateToProps(state) {
  return state.newApp;
}

export function mapDispatchToProps(dispatch) {
  return {
    handleSubmit: app => dispatch(postNewApp(app)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewAppPage);
