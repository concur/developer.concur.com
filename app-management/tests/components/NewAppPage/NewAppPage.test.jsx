import React from 'react';
import { shallow } from 'enzyme';
import { createStore } from 'redux';
import { shallowToJson } from 'enzyme-to-json';

import NewAppPage from '../../../components/NewAppPage/NewAppPage';
import { mapStateToProps, mapDispatchToProps } from '../../../components/NewAppPage';
import NewAppForm from '../../../components/NewAppPage/NewAppForm';
import appReducer from '../../../reducers';

const store = createStore(appReducer);

describe('<NewAppPage />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render with <NewAppForm />', () => {
    const props = {
      handleSubmit: jest.fn(),
      isFetching: false,
      error: '',
    };
    const page = shallow(<NewAppPage {...props} />);

    expect(shallowToJson(page)).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('returns entire newApp state', () => {
      const state = mapStateToProps(store.getState());

      expect(state).toEqual(store.getState().newApp);
    });
  });

  describe('mapDispatchToProps', () => {
    it('returns submit handler', () => {
      const dispatcher = jest.fn();
      const state = mapDispatchToProps(dispatcher);
      state.handleSubmit('app');

      expect(state.handleSubmit).toBeDefined();
      expect(dispatcher).toHaveBeenCalled();
    });
  });
});
