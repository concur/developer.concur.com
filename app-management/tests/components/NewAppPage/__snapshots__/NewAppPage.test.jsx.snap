exports[`<NewAppPage /> should render with <NewAppForm /> 1`] = `
<div
  className="row">
  <div
    className="col-md-12">
    <h2>
      Create an App
    </h2>
    <LoadingSpinner
      loading={false}
      message="Submitting..." />
    <ErrorAlert
      error="" />
    <div
      className="alert alert-info">
      <p>
        Register your new application (app) with Concur by filling out this form. Once you have registered an app, you will receive a clientId and clientSecret. The clientId is a unique UUID4 identifier for your app, and the clientSecret is your appâ€™s password. You will be using this credential to obtain tokens either for the app itself, or on behalf of a user.
      </p>
      <p>
        For more information on this process, please refer to our Getting Started guide for
        <a
          href="/api-reference/authentication/getting-started.html"
          rel="noopener noreferrer"
          target="_blank">
           authentication
        </a>
      </p>
    </div>
    <ReduxForm
      onSubmit={[Function]} />
  </div>
</div>
`;
